### ISSUE TITLE
How to look at the loss value?

### ISSUE THREAD

**Issue by asanakoy** [lmcinnes/umap#100]
Is there a possibility to print the loss value during the optimization process? Or at least after the last epoch?

It would make sense to add this feature, as it would help to set an appropriate number of epochs when one tunes the algorithm for specific data.

**Comment by vseledkin** [comment]
Probably this wont be easy because of negative sampling, it is endless sourse of gradient proportional to current learning rate amplitude

**Comment by lmcinnes** [comment]
@vseledkin is correct, the negative sampling is a method used to avoid ever actually computing the full loss value which is remarkably expensive. I did have a function that could be used to compute loss, but in practice it only scaled to datasets of a few thousand points, so I never included it.

**Comment by asanakoy** [comment]
@lmcinnes but how can we estimate that the method has converged?

**Comment by lmcinnes** [comment]
Convergence is not checked - -instead the optimization is run fora
specified number of epochs.

**Comment by asanakoy** [comment]
But how can I make sure that 300 epoch is not better than 200 epochs? 
If we had a loss function or at least the gradient norms, than we could infer how many epochs is enough.

**Comment by sleighsoft** [comment]
Contributions welcome :)

**Comment by AndLen** [comment]
> @vseledkin is correct, the negative sampling is a method used to avoid ever actually computing the full loss value which is remarkably expensive. I did have a function that could be used to compute loss, but in practice it only scaled to datasets of a few thousand points, so I never included it.

I don't suppose you would be willing to share it? I'm looking at applications of UMAP on reasonably small data, where having the loss function explicitly is every useful.

**Comment by csinva** [comment]
+1

**Comment by timsainb** [comment]
Parametric UMAP has a non-parametric module that saves loss

https://github.com/lmcinnes/umap/blob/master/notebooks/Parametric_UMAP/06.0-nonparametric-umap.ipynb
